### PR TITLE
#10074 switch to hilling/cdi-test

### LINKED ISSUE #10074: Replace CDI unit testing framework
### Problem Description
Unit testing of the SORMAS backend is currently facing multiple problems:

* Bean-test no longer maintained (e.g. no support for jakarta-ee, using very old versions of weld + deltaspike) -> we would have to fork and maintain on our own
* Combination of bean-test and mocking is not working very good. E.g. it's tricky to change the logged-in user during test run and hard to mock configuration properties, feature configurations
* Transaction logic is partly not testable (e.g. annotation TransactionAttribute)
* Validation annotations are not testable (e.g. #7611)
* Parts of logic data stored in JSON format can't be tested (e.g. properties of feature configurations). -> This will not be addressed here. Maybe an update of H2 can also fix this problem.
* Is overall a bit slow, because we are setting up the complete backend CDI container for each test class

#### Possible Test Approaches

Generally speaking there are three approaches to testing backend functionality:

1. Basic unit tests + mocking
* Mocking all the dependencies of tested services and facades would be a lot of work and make the tests a lot less helpful
2. Integration tests (e.g. based on Arquillian)
* Every test run requires an archive to be built and deployed which makes the testing slower. Too slow if you need to run thousands of test cases.
* Creating the deployments, especially resolving the required artifacts and classes, is not trivial.
* Setup tends to be complicated.
* Still helpful **for functionality that can't be tested otherwise** (e.g. transaction based stuff or database specific stuff)
3. CDI-based testing (provides bean container that automatically injects dependencies)
* This is the approach we have followed so far and **should go on using**.
* It's currently facing the problems mentioned above, though

### Proposed Change

- [ ] Find a replacement for the novatec bean-test framework.

* cdi-test:  Small open source project, but maintained, using a current version of weld and is not relying on deltaspike (-> faster)
  It's supposed to make mocking of parts of the system a lot easier
* cdi-unit: http://cdi-unit.github.io/cdi-unit/
* weld-testing: https://dzone.com/articles/weld-junit-easy-testing-of-cdi-beans

- [ ] Check whether both can be used in parallel to simplify migration
- [ ] Make transaction logic testable, e.g. provide an interceptor for TransactionAttribute annotation
- [ ] Make validation testable
- [ ] Create implementation/migration issue(s)

### Additional Information

Alternative approach:Use Spring instead of JavaEE for the backend. This would probably be a very big effort. If we are considering this it would have to be evaluated. 

https://cdi-test.hilling.de/
http://www.mastertheboss.com/java-ee/jakarta-ee/testing-jakarta-ee-applications-with-cdi-test/
https://entwickler.de/container/qualitatssicherung-im-container-leichtgewichtiges-testen-von-cdi-komponenten
https://www.sigs-datacom.de/uploads/tx_dmjournals/barragan_wanner_baier_JS_05_13_kjgu.pdf
https://antoniogoncalves.org/2018/01/16/java-ee-vs-spring-testing/
https://itnext.io/testing-jakarta-ee-9-applications-with-arquillian-and-payara-6-52fd153d8d9
https://medium.com/riag-digital-techblog/why-you-should-migrate-to-junit-5-15438f5c56f

### PR DESCRIPTION
**Pull request by StefanKock**
Fixes #10074.

Incomplete, see https://github.com/hzi-braunschweig/SORMAS-Project/issues/10074#issuecomment-1418861485

### CODE REVIEW

File: sormas-backend/src/test/resources/META-INF/persistence.xml [hunk 1]
```diff
@@ -100,7 +100,8 @@
 	    	<property name="javax.persistence.jdbc.user" value="sa" />
 	    	<property name="javax.persistence.jdbc.password" value="" />
 
-      		<property name="hibernate.hbm2ddl.auto" value="create-drop" />
+			<!-- Database schema is created before each test by TestDatabaseCleaner class -->
+      		<property name="hibernate.hbm2ddl.auto" value="none" />
```
Comment (StefanKock) [inline sormas-backend/src/test/resources/META-INF/persistence.xml:104]: Remove the config, it's default value

File: sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+	protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+	public void run(Connection connection) throws SQLException {
+
+		connection.prepareStatement("DROP ALL OBJECTS").execute();
+		connection.commit();
+
+		Map<String, String> properties = new HashMap<>();
+		properties.put("hibernate.hbm2ddl.auto", "create");
```
Comment (StefanKock) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java:32]: Use constants for better documentation (I used the `jakarta.persistence.schema-generation.database.action` definition here to be closer to JPA standard):

```
import org.hibernate.cfg.AvailableSettings;
import org.hibernate.tool.schema.Action;

		properties.put(AvailableSettings.HBM2DDL_DATABASE_ACTION, Action.CREATE.getExternalJpaName());
```

File: sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ * </p>
+ */
+public class TestDatabaseCleaner implements DatabaseCleaner {
+
+	protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+	public void run(Connection connection) throws SQLException {
+
+		connection.prepareStatement("DROP ALL OBJECTS").execute();
+		connection.commit();
```
Comment (StefanKock) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java:29]: Superfluous: Drop will be done in the next step anyway.

**Review by StefanKock — COMMENTED** [review]
I reviewed the database reset

File: sormas-backend/pom.xml [hunk 4]
```diff
@@ -121,20 +121,28 @@
...
-		<dependency>
-			<groupId>org.apache.deltaspike.core</groupId>
-			<artifactId>deltaspike-core-impl</artifactId>
+			<groupId>de.hilling.junit.cdi</groupId>
+			<artifactId>cdi-test-core</artifactId>
+			<version>3.4.0</version>
+			<scope>test</scope>
+			<exclusions>
+				<exclusion>
+					<artifactId>junit-jupiter-api</artifactId>
```
Comment (StefanKock) [inline sormas-backend/pom.xml:130]: Do not exclude `<artifactId>junit-jupiter-api</artifactId>` here, but add it in the dependency management in the sormas-base/pom.xml to override the version,

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11443

**Comment by MartinWahnschaffe** [comment]
@StefanKock Ready to be reviewed. Github SORMAS CI now takes 30 minutes instead of 70 minutes. Running the backend unit test is done from 45 minutes to 15 minutes on my old system and down to 10 minutes on my new one.

These are the things I consider out of scope and would move into a separate issue:

- Refactor creation of users + login
- Encapsulate MockProducer.getPrincipal 
- Reformat & organize imports of all unit test files

As a next step I will create a short readme for the backend unit tests that sums up the most important things to understand. I consider this quite important and would like to include it as a part of this issue or at least have it ready when merging the pull request.

File: sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java (hunk 2 above)
Comment (MartinWahnschaffe) [reply sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java]: Done. Except getExternalJpaName is not available on our current hibernate version.

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11443

File: sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java (hunk 2 above)
Comment (StefanKock) [reply sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java]: Okay, #11032 brings this version update: `org.hibernate:hibernate-core ............. 5.6.5.Final -> 5.6.14.Final`

File: sormas-backend/src/test/java/de/symeda/sormas/backend/AbstractBeanTest.java [hunk 5]
```diff
@@ -855,13 +851,13 @@ protected UserDto useNationalAdminLogin() {
...
 			creator.getUserRoleReference(DefaultUserRole.ADMIN));
-		when(MockProducer.getPrincipal().getName()).thenReturn("NationalAdmin");
-
+		loginWith(natUser);
 		return natUser;
 	}
 
 	protected void loginWith(UserDto user) {
 		when(MockProducer.getPrincipal().getName()).thenReturn(user.getUserName());
+		getCurrentUserService().getCurrentUser(); // load into cache, to work-around CurrentUserService.fetchUser @Transactional not working in tests
```
Comment (StefanKock) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/AbstractBeanTest.java]: No comment in the same line with code, comment before

File: sormas-backend/src/test/java/de/symeda/sormas/backend/AbstractBeanTest.java [hunk 6]
```diff
@@ -857,7 +864,7 @@ protected UserDto useNationalAdminLogin() {
 
 	protected void loginWith(UserDto user) {
 		when(MockProducer.getPrincipal().getName()).thenReturn(user.getUserName());
-		getCurrentUserService().getCurrentUser(); // load into cache, to work-around CurrentUserService.fetchUser @Transactional not working in tests
+		getCurrentUserService().getCurrentUser(); // load into cache, to work-around CurrentUserService.fetchUser @TransactionalCustom not working in tests
```
Comment (StefanKock) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/AbstractBeanTest.java:860]: Comment above code

File: sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java [hunk 7]
```diff
@@ -25,12 +26,12 @@ public class TestDatabaseCleaner implements DatabaseCleaner {
...
 	public void run(Connection connection) throws SQLException {
 
+		// needed to drop h2 functions
 		connection.prepareStatement("DROP ALL OBJECTS").execute();
 		connection.commit();
 
 		Map<String, String> properties = new HashMap<>();
-		properties.put("hibernate.hbm2ddl.auto", "create");
-		Persistence.generateSchema("beanTestPU", properties);
+		properties.put(AvailableSettings.HBM2DDL_DATABASE_ACTION, "create");//Action.CREATE.getExternalJpaName());
```
Comment (StefanKock) [inline sormas-backend/src/test/java/de/symeda/sormas/backend/TestDatabaseCleaner.java]: Add marker `// TODO #11032 Switch to constant from org.hibernate.tool.schema.Action`

File: sormas-backend/src/test/resources/META-INF/persistence.xml [hunk 8]
```diff
@@ -101,7 +103,7 @@
 	    	<property name="javax.persistence.jdbc.password" value="" />
 
 			<!-- Database schema is created before each test by TestDatabaseCleaner class -->
-      		<property name="hibernate.hbm2ddl.auto" value="none" />
+      		<!--<property name="hibernate.hbm2ddl.auto" value="create"/>-->
```
Comment (StefanKock) [inline sormas-backend/src/test/resources/META-INF/persistence.xml:106]: Replace code by `<!-- Database schema is freshly created before each test by TestDatabaseCleaner -->`

File: sormas-base/pom.xml [hunk 9]
```diff
@@ -868,6 +868,19 @@
 				<scope>test</scope>
 			</dependency>
 
+			<dependency>
+				<groupId>de.hilling.junit.cdi</groupId>
+				<artifactId>cdi-test-core</artifactId>
+				<version>3.4.0</version>
+				<scope>test</scope>
+				<exclusions>
+					<exclusion>
```
Comment (StefanKock) [inline sormas-base/pom.xml:877]: Don't exclude it, list the `junit-jupiter-api` also in the `dependencyManagement` with the desired junit5 version

File: sormas-backend/pom.xml [hunk 10]
```diff
@@ -121,20 +121,9 @@
...
-		<dependency>
-			<groupId>org.apache.deltaspike.core</groupId>
-			<artifactId>deltaspike-core-impl</artifactId>
-		</dependency>
-		<dependency>
-			<groupId>org.apache.deltaspike.cdictrl</groupId>
-			<artifactId>deltaspike-cdictrl-weld</artifactId>
+			<groupId>de.hilling.junit.cdi</groupId>
+			<artifactId>cdi-test-core</artifactId>
+			<scope>test</scope>
```
Comment (StefanKock) [inline sormas-backend/pom.xml:126]: scope test to be defined in sormas-base/pom.xml

**Review by StefanKock — COMMENTED** [review]
I reviewed the code (no time for full test runs), and only have some comment remarks and POM changes.

**Comment by sormas-vitagroup** [comment]
SonarCloud analysis: https://sonarcloud.io/dashboard?id=SORMAS-Project&pullRequest=11443

**Pull request merged**